### PR TITLE
Mutation Conventions

### PR DESCRIPTION
**Pull request by michaelstaib**
- [ ] Ensure that conventions play nice with the relay query field rule.
- [x] Code-First Tests
- [x] Annotation-Based Tests
- [x] Schema-First
  - [ ] Schema-First Directive `@mutation`
- [x] Ensure #4517 is resolved

### CODE REVIEW

**Comment by michaelstaib** [comment]
I have reworked how we handle arguments so that all schema-building approaches are aligned. With the new approach, we rewrite the arguments in field middleware. This allows us to have the resolver code act if there was no input type at all.
This will cause allocation for two dictionaries for each mutation field. Since in general, we will not have more than a couple of mutation fields per request this should be OK. We can always rent these if the need would arise.

File: src/HotChocolate/Core/src/Execution/Processing/ArgumentValue.cs [hunk 1]
```diff
@@ -4,17 +4,19 @@
 
 namespace HotChocolate.Execution.Processing;
 
-public sealed class ArgumentValue
+public sealed class ArgumentValue : IInputFieldInfo
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Execution/Processing/ArgumentValue.cs:7]: The `ArgumentValue` is now a `IInputFieldInfo`.

The `IInputFieldInfo` interface provides access to the most important properties of an input field. With this, we can now substitute an actual field with a made-up field in core helpers like the `InputParser`. 

In our concrete case, we can preserve the resolver arguments allowing us to rewrite the schema without the need for the resolver aligning to this new schema. The resolver in this case can still act against the original schema definition.

File: src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
+namespace HotChocolate.Types;
+
+public class MutationAttribute : ObjectFieldDescriptorAttribute
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs:3]: XML Documentation

File: src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    /// The name of the field in the payload
+    /// <code>
+    /// type CreateUserPayload {
+    ///    thisIsTheFieldName: User
+    /// }
+    /// </code>
+    /// </summary>
+    public string? PayloadFieldName { get; set; }
+
+    public bool Enabled { get; set; } = true;
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs:45]: Still not happy with this property name.

File: src/HotChocolate/Core/src/Types.Mutations/MutationConventionOptions.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+namespace HotChocolate.Types;
+
+public struct MutationConventionOptions
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/MutationConventionOptions.cs:3]: XML Documentation

File: src/HotChocolate/Core/src/Types.Mutations/MutationMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        _inputArgumentName = inputArgumentName;
+        _resolverArguments = resolverArguments;
+    }
+
+    public ValueTask InvokeAsync(IMiddlewareContext context)
+        => InvokeInternalAsync(((MiddlewareContext)context));
+
+    private async ValueTask InvokeInternalAsync(MiddlewareContext context)
+    {
+        var input = context.ArgumentValue<IDictionary<string, object?>>(_inputArgumentName);
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/MutationMiddleware.cs:29]: We need documentation for this

File: src/HotChocolate/Core/src/Types.Mutations/ResolverArgument.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#nullable enable
+
+namespace HotChocolate.Types;
+
+internal sealed class ResolverArgument : IInputFieldInfo
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types.Mutations/ResolverArgument.cs:5]: This represents our resolver argument substitute which we preserve in our mutation middleware.

File: src/HotChocolate/Core/src/Types/Configuration/AggregateTypeInterceptor.cs [hunk 7]
```diff
@@ -54,11 +55,19 @@ public void SetInterceptors(IReadOnlyCollection<object> interceptors)
...
         IDescriptorContext context,
+        TypeInitializer typeInitializer,
+        TypeRegistry typeRegistry,
+        TypeLookup typeLookup,
         TypeReferenceResolver typeReferenceResolver)
     {
         foreach (TypeInterceptor interceptor in _typeInterceptors)
         {
-            interceptor.InitializeContext(context, typeReferenceResolver);
+            interceptor.InitializeContext(
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Configuration/AggregateTypeInterceptor.cs:65]: We need to have a look at this again ... do we need all of these internal services?

File: src/HotChocolate/Core/src/Types/Resolvers/DefaultResolverCompiler.cs [hunk 8]
```diff
@@ -448,7 +448,8 @@ private static bool IsPureResolverResult(Type resultType)
         ParameterInfo parameter,
         IParameterExpressionBuilder[] fieldParameterExpressionBuilders)
     {
-        if (_cache.TryGetValue(parameter, out IParameterExpressionBuilder? cached))
+        if (fieldParameterExpressionBuilders.Length == 0 &&
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Resolvers/DefaultResolverCompiler.cs:451]: This allows us to recompile a resolver by overwriting the expression builders with locally provided ones.

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Resolvers/DefaultResolverCompiler.cs:451]: We might want to ensure that local field expression builder are not cached.

File: src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs (hunk 3 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs:45]: Disabled = false?

Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs:45]: or maybe even [DisableMutationConvetion]

File: src/HotChocolate/Core/src/Execution/Processing/ArgumentValue.cs (hunk 1 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Execution/Processing/ArgumentValue.cs:7]: Can we put this inside the code with some documentation?

File: src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs (hunk 3 above)
Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types.Mutations/MutationAttribute.cs]: I changed it to `Disable`.

**Pull request merged**